keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,404 @@
+
+After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.
+
+Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:
+
+- pods that are backed by a provisioner that support safe detach
+- pods that are backed by volumes that aren't marked as `ReadWriteMany, this would avoid breaking having multiple replicas of the same pod.
+
+Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `SAFE_DETACH` CSI controller capability is true. If true, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.
+
+This would trigger the deletion of the `volumeAttachement` objects, which would in turn trigger `ControllerUnpublishVolume()`. The provider would need to ensure there that no volume is being used through a mount point. If unable to determine the usage, the storage provider can return an error without detaching the volume, the `external-attacher` would then retry with backoff.
+
+The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due the shutdown.
+
+### Handle the Return of Shutdown Node
+
+If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.
+
+The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.
+
+Kubelet should check the `SAFE_DETACH` node service capability. If it is true, it should forcefully unmount the volumes stuck on the node by calling `NodeUnpublishVolume` and/or `NodeUnstageVolume`.
+
+The Pod GC Controller should also remove the `quarantine` taint from the nodes that are already cleaned up and are safe to run pods again.
+
+### Risks and Mitigations
+
+This KEP introduces changes, and if not tested correctly could result in data corruption for users.
+To mitigate this we plan to have a high test coverage and to introduce this enhancement as an alpha
+
+## Design Details
+
+### Test Plan
+
+This feature will be tested with the following approaches:
+
+- unit tests
+- integration
+- e2e
+
```

Comment:
This is not a great test plan, please expand on each test to add details, thanks!